ENH: Improve the ARM cpu feature detection by parsing /proc/cpuinfo

Qiyu8 opened the pull request:
According to industry best practices, cpu features should be fetched from both hwcaps(`getauxval`, `/proc/self/auxv`) and `/proc/cpuinfo` so we can have some information if the executable is sandboxed.Also resolves the issue of badly patched kernels, usually the old arm7-a kernels on Andriod devices don't provide support for auxiliary vector or sometimes selinux prevent reading auxiliary vector (/proc/self/auxv) on non-root users.

Reference:

1. [Google CPU Detection.](https://github.com/google/cpu_features/blob/master/src/cpuinfo_arm.c)
2. [OpenCV CPU Detection.](https://github.com/opencv/opencv/blob/master/3rdparty/cpufeatures/cpu-features.c)
3. [Runtime detection of CPU features on an ARMv8-A CPU.](https://community.arm.com/developer/tools-software/oss-platforms/b/android-blog/posts/runtime-detection-of-cpu-features-on-an-armv8-a-cpu)
4. [Fix detecting and testing armhf features](https://github.com/numpy/numpy/pull/16304)

Qiyu8 committed:
parse /proc/cpuinfo

vrakesh commented:
I tested this on an EC2 ARM instance, as well this works well. awesome we are improving this

seiko2plus commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -331,8 +518,7 @@ npy__cpu_init_features_linux(void)
         /*
-         * FIXME: failback to compiler definitions,
-         * BTW we can parse /proc/cpuinfo for badly patched kernels
+         * failback to compiler definitions
```
No need for duplicate parsing.  You can fill `hwcap` and `hwcap2` bits from `/proc/cpuinfo` in here, only when parsing `auxv` fail.

Qiyu8 replied:
reordered to getauxval  --> auxv --> cpuinfo.

Qiyu8 committed:
when parsing auxv fail, try parsing /proc/cpuinfo

charris commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -298,6 +301,188 @@ npy__cpu_init_features_arm8(void)
+                continue;
+            break;
+        }
+        if (ret == 0)
```
Always use `{...}`, even for single line statements.

charris commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -298,6 +301,188 @@ npy__cpu_init_features_arm8(void)
+                count = -1;
+            break;
+        }
+        if (ret == 0)
```
See above.

charris commented:
Lots of C style nits. Please follow the style guide `doc/C_STYLE_GUIDE.rst.txt`.

Qiyu8 committed:
modify according to NumPy C Style Guide

Qiyu8 committed:
Update npy_cpu_features.c.src

Qiyu8 committed:
Update npy_cpu_features.c.src

mattip commented:
Is there an open issue this fixes? 

Did you consider to call out to python to do the parsing rather than writing it in C? Then you could use a line-based parser rather than having to allocate a buffer for the whole file, which requires you to first determine the length of the buffer. That call could return the `flags` line. For an example, see the use of [`numpy.core._dtype`](https://github.com/numpy/numpy/blob/v1.19.0rc1/numpy/core/src/multiarray/descriptor.c#L64)

If you do choose to keep this implementation, it looks very similar to the [OpenCV one](https://github.com/opencv/opencv/blob/master/3rdparty/cpufeatures/cpu-features.c). That code is copyright, we should be careful to respect the license at the top of the file, which means it is fine to use it but we should give credit by following the instructions there.

Qiyu8 commented:
@mattip Thanks for your thoughtful comment, I will investigate into the call-python parse way, or following the copyright License if not succeed.

Qiyu8 committed:
Merge branch 'master' of github.com:numpy/numpy into improve-arm-cpudetect

Qiyu8 committed:
call-python way to get cpuinfo

Qiyu8 committed:
call-python way to get cpuinfo

seiko2plus commented:
@Qiyu8, ARM chips are cute and simple please don't torture it, python already doing a great job on that. Bringing a python module to C world isn't something cheap is worth than the previous code with a dozen of times.
My suggestion to you is to keep your previous code and leave the notice about the copyright, later we can replace it with on-fly parsing, also put into your consideration that `/proc/cpuinfo` normally mapped to a region in memory, that's why define `/proc`  as a "virtual filesystem".
Thanks!.

Qiyu8 committed:
call-python way to get cpuinfo

mattip commented:
> Bringing a python module to C world isn't something cheap, it's worse than the previous code with a dozen of times.

Sorry, that was my suggestion. It is worth benchmarking how much this slows down import. There are other places we do the same trick (for converting dtypes to ctypes and back, and more) so if it is significant we may want to rethink that strategy.

Qiyu8 committed:
compare C way and call-python way to get cpuinfo

Qiyu8 commented:
@mattip Comparing to C way, although the code size has decreased about 50%, the performance gap is significant, here is the benchmark result(per clock):
```
C way time : 2111
call-python way time : 31047
```
Should I keep previous code like @seiko2plus said,  or use the call-python code that easier to maintain?

Qiyu8 committed:
compare C way and call-python way to get cpuinfo

mattip commented:
What units is that in? On my raspberrypi4 with a mmc flash memory, `time python3 -c "import os"` takes 75ms, and `time python3 -c "import numpy"` takes 356ms (for the apt-provided 1.16.2 on python3). So for you this would take 15x more or ~6 seconds to import? If so it seems the python code is too much of a performance hit and we should prefer the C code, although if the machine is so slow I wonder what processing it could possibly do.

Qiyu8 committed:
Keep the C way.

Qiyu8 commented:
I'm using clock() function from time.h, Here is an example:
```
#include<stdio.h>
#include<time.h>
clock_t start, stop;
double duration;
int main()
{
    start=clock();
    //******************************
    //*Here is the program you want to test the performance* 
    //******************************
    stop=clock();
    duration=(double)(stop-start)/CLK_TCK;
    printf("%f\n",duration);
    return 0;
}
```

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -1,3 +1,30 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
```
I think this license should only apply to code actually taken from outside sources. Could you split this into two files: one with the code from Android, and another with everything else. Then the second file could simply `#include` the first.

Qiyu8 committed:
update
Merge branch 'master' of github.com:Qiyu8/numpy into improve-arm-cpudetect

Qiyu8 committed:
split into two files because of the copyright.

Qiyu8 committed:
Update npy_cpuinfo_parser.h

mattip commented:
LGTM. Waiting for CI. It seems complicated to set up a test for this: we would need to introduce an environment variable or somehow cause "normal" detection to fail.

Qiyu8 commented:
yes, It's hard to test an hardware&software environment-related situation(that's why it's called back-up plan 😄 ), but this feature is easy to test locally,  all you need to do is copy `get_feature_from_proc_cpuinfo` into the first line of `npy__cpu_init_features_linux`.

Qiyu8 committed:
update
Merge branch 'master' of https://github.com/numpy/numpy into improve-arm-cpudetect

Qiyu8 committed:
only set hwcap in order to improve the program's robustness

Qiyu8 committed:
update
Merge branch 'improve-arm-cpudetect' of github.com:Qiyu8/numpy into improve-arm-cpudetect

Qiyu8 committed:
improve performance.

mattip commented:
Close/reopen to get latest master changes

mattip commented:
Thanks @Qiyu8
